code: added Typescript and Javascript examples

RiverTwilight opened the pull request:
Not sure whether these formats meet the requirement. If everything is okay I will continue to transcribe more:-)

If this PR is related to coding or code translation, please fill out the checklist.

- [x] I've tested the code and ensured the outputs are the same as the outputs of reference codes.
- [x] I've checked the codes (formatting, comments, indentation, file header, etc) carefully.
- [x] The code does not rely on a particular environment or IDE and can be executed on a standard system (Win, macOS, Ubuntu).

RiverTwilight commented:
I've finished transcribing `time_complexity.js` and some codes in the doc, could you review these changes? I will continue work on the typescript if everything goes well. @krahets 

Code run result here:
![image](https://user-images.githubusercontent.com/52880665/210224608-ebd9b8aa-579d-4176-8c70-2c16a848091d.png)

krahets commented on docs/chapter_computational_complexity/time_complexity.md:
```diff
@@ -1935,8 +1983,8 @@ $$
 
-- 当 `nums = [?, ?, ..., 1]`，即当末尾元素是 $1$ 时，则需完整遍历数组，此时达到 **最差时间复杂度 $O(n)$** ；
-- 当 `nums = [1, ?, ?, ...]` ，即当首个数字为 $1$ 时，无论数组多长都不需要继续遍历，此时达到 **最佳时间复杂度 $\Omega(1)$** ；
+-   当 `nums = [?, ?, ..., 1]`，即当末尾元素是 $1$ 时，则需完整遍历数组，此时达到 **最差时间复杂度 $O(n)$** ；
```
Please revert this line back to the original version.

krahets commented on docs/chapter_computational_complexity/time_complexity.md:
```diff
@@ -399,19 +399,39 @@ $$
+        a *= 2; // +1
+        // 循环 n 次
+        for(var i = 0; i < n; i++){ // +1
+            console.log(0) // +1
```
`;` at the end of every line.

krahets commented on docs/chapter_computational_complexity/time_complexity.md:
```diff
@@ -148,9 +148,9 @@ $$
-- 算法 `A` 只有 $1$ 个打印操作，算法运行时间不随着 $n$ 增大而增长。我们称此算法的时间复杂度为「常数阶」。
-- 算法 `B` 中的打印操作需要循环 $n$ 次，算法运行时间随着 $n$ 增大成线性增长。此算法的时间复杂度被称为「线性阶」。
-- 算法 `C` 中的打印操作需要循环 $1000000$ 次，但运行时间仍与输入数据大小 $n$ 无关。因此 `C` 的时间复杂度和 `A` 相同，仍为「常数阶」。
+-   算法 `A` 只有 $1$ 个打印操作，算法运行时间不随着 $n$ 增大而增长。我们称此算法的时间复杂度为「常数阶」。
```
Please revert this line back to the original version.

krahets commented on codes/typescript/chapter_computational_complexity/time_complexity.ts:
```diff
@@ -0,0 +1,5 @@
+ * File: time_complexity.js
+ * Created Time: 2023-01-02
+ * Author: RiverTwilight (contact@rene.wang)
+ */
```
Do you forget commit the code?

krahets commented on codes/javascript/chapter_computational_complexity/time_complexity.js:
```diff
@@ -0,0 +1,159 @@
+ * Author: RiverTwilight (contact@rene.wang)
+ */
+
+class time_complexity {
```
We don't have to create a class for these functions in JS and TS. Please refer to other JS and TS codes to see the details.

RiverTwilight replied:
So sorry for this. I made this decision as the Java code does. Because the contribution guide recommended doing so.

krahets commented on codes/javascript/chapter_computational_complexity/time_complexity.js:
```diff
@@ -0,0 +1,159 @@
+	}
+}
+
+(function main() {
```
Feel free to delete the main function. (Refer to other JS and TS codes.)

krahets commented on codes/javascript/chapter_computational_complexity/time_complexity.js:
```diff
@@ -0,0 +1,159 @@
+class time_complexity {
+	/* 常数阶 */
+	constant(n) {
+		var count = 0;
```
`let` is preferrable.

RiverTwilight replied:
In many cases, the count is dynamic (for example, excuting`count++`) so I choose `var`. As for let or var, I only use `var` when the variable is at the top level of a function but actually a personal habit I can of course change them all here:D
![image](https://user-images.githubusercontent.com/52880665/210230891-1aedaf22-04bf-4f70-a69a-5dd5aeb48bc1.png)

krahets replied:
More information at [understanding-let-const-and-var-keywords](https://www.freecodecamp.org/news/understanding-let-const-and-var-keywords/).

In short:

That's all, my friends. You need to consider these circumstances and concepts to evaluate how var, let, and const behave. So, the rule goes:

Don't use var anymore.
Use let or const.
Use const more often. Use let when you need to reassign another value to a variable.
Don't try to access a variable without declaring it.

krahets commented on codes/javascript/chapter_computational_complexity/time_complexity.js:
```diff
@@ -0,0 +1,159 @@
+	constant(n) {
+		var count = 0;
+		const size = 100000;
+		for (var i = 0; i < size; i++) count++;
```
We use "four spaces" as the indentation, not "tab", for the JS and TS code.

RiverTwilight replied:
So I recommend to add a prettier config in the root directory:D

krahets replied:
Feel free to add it~

krahets reviewed (commented):
Thanks! Please address the comments.

RiverTwilight commented:
All addressed @krahets

krahets commented on codes/typescript/chapter_computational_complexity/time_complexity.ts:
```diff
@@ -0,0 +1,153 @@
+/* 常数阶 */
+function constant(n: number): number {
+	let count = 0;
+	const size = 100000;
```
Replace tabs with four spaces.

krahets commented on codes/javascript/chapter_computational_complexity/time_complexity.js:
```diff
@@ -0,0 +1,153 @@
+/* 常数阶 */
+function constant(n) {
+	let count = 0;
+	const size = 100000;
```
Replace tabs with four spaces.

krahets reviewed (commented):
Some unaddressed comments are found.

RiverTwilight commented:
> Some unaddressed comments are found.

Here

krahets reviewed (commented):
Several unaddressed comments in `time_complexity.md`.

RiverTwilight commented:
> Several unaddressed comments in `time_complexity.md`.

Oh, I couldn't catch it anymore...as above mentioned I didn't change any text. You may need to check the raw file instead of the code diff.

![image](https://user-images.githubusercontent.com/52880665/210238800-92d2efa6-6e9c-4421-b386-72488b28e941.png)

![image](https://user-images.githubusercontent.com/52880665/210238844-3267d590-2259-4012-b9d2-1c17541275cc.png)

krahets commented:
<img width="1027" alt="image" src="https://user-images.githubusercontent.com/26993056/210239211-1b183209-011c-45a6-9f67-d80b3a84395e.png">

There should be only one space between the `-` and the contents. You've modified it to more than one space. Please revert them.

RiverTwilight commented:
> <img alt="image" width="1027" src="https://user-images.githubusercontent.com/26993056/210239211-1b183209-011c-45a6-9f67-d80b3a84395e.png">
> 
> There should be only one space between the `-` and the contents. You've modified it to more than one space. Please revert them.
> 
> You can clearly catch the differences on this page.
> 
> <img alt="image" width="926" src="https://user-images.githubusercontent.com/26993056/210239441-4ef10407-9cbc-4f7c-aa5b-cf08baaaac99.png">

OMG, it's my formatter dumming...reeeally sorry for that...

krahets commented on docs/chapter_computational_complexity/time_complexity.md:
```diff
@@ -354,19 +399,39 @@ $$
+        }
+
+    }
 
```
Please remove the empty line. Same for others.

krahets reviewed (commented):
One more comment, please.

justin-tse commented on codes/typescript/chapter_computational_complexity/time_complexity.ts:
```diff
@@ -0,0 +1,153 @@
+}
+
+/* 线性阶（遍历数组） */
+function arrayTraversal(nums: number[]) {
```
```suggestion
function arrayTraversal(nums: number[]): number {
```

justin-tse commented on codes/typescript/chapter_computational_complexity/time_complexity.ts:
```diff
@@ -0,0 +1,153 @@
+    return count;
+}
+
+var n = 8;
```
```suggestion
/* Driver Code */
// 可以修改 n 运行，体会一下各种复杂度的操作数量变化趋势
const n = 8;
```

justin-tse commented on codes/typescript/chapter_computational_complexity/time_complexity.ts:
```diff
@@ -0,0 +1,153 @@
+var n = 8;
+console.log("输入数据大小 n = " + n);
+
+var count = constant(n);
```
```suggestion
let count = constant(n);
```

justin-tse commented on docs/chapter_computational_complexity/time_complexity.md:
````diff
@@ -79,13 +79,27 @@ $$
 === "JavaScript"
 
     ```js title=""
````
```suggestion
    ```js title=""
    // 在某运行平台下
```

justin-tse commented on docs/chapter_computational_complexity/time_complexity.md:
```diff
@@ -79,13 +79,27 @@ $$
+    function algorithm(n) {
+        var a = 2; // 1 ns
+        a = a + 1; // 1 ns
+        a = a * 2; // 10 ns
```
```suggestion
        a = a * 2; // 10 ns
        // 循环 n 次
```

justin-tse commented on docs/chapter_computational_complexity/time_complexity.md:
````diff
@@ -79,13 +79,27 @@ $$
 
 === "TypeScript"
 
     ```typescript title=""
````
```suggestion
    ```typescript title=""
```
```suggestion
    ```typescript title=""
    // 在某运行平台下
```

justin-tse commented on docs/chapter_computational_complexity/time_complexity.md:
```diff
@@ -79,13 +79,27 @@ $$
+    function algorithm(n: number): void {
+        var a: number = 2; // 1 ns
+        a = a + 1; // 1 ns
+        a = a * 2; // 10 ns
```
```suggestion
        a = a * 2; // 10 ns
        // 循环 n 次
```

justin-tse commented on docs/chapter_computational_complexity/time_complexity.md:
```diff
@@ -79,13 +79,27 @@ $$
+        var a: number = 2; // 1 ns
+        a = a + 1; // 1 ns
+        a = a * 2; // 10 ns
+        for(var i = 0; i < n; i++) { // 1 ns ，每轮都要执行 i++
```
```suggestion
        for(let i = 0; i < n; i++) { // 1 ns ，每轮都要执行 i++
```

justin-tse commented on docs/chapter_computational_complexity/time_complexity.md:
```diff
@@ -79,13 +79,27 @@ $$
+        var a = 2; // 1 ns
+        a = a + 1; // 1 ns
+        a = a * 2; // 10 ns
+        for(var i = 0; i < n; i++) { // 1 ns ，每轮都要执行 i++
```
```suggestion
        for(let i = 0; i < n; i++) { // 1 ns ，每轮都要执行 i++
```

justin-tse commented on docs/chapter_computational_complexity/time_complexity.md:
```diff
@@ -220,13 +234,44 @@ $$
+    }
+    // 算法 B 时间复杂度：线性阶
+    function algorithm_B(n) {
+        for (var i = 0; i < n; i++) {
```
```suggestion
        for (let i = 0; i < n; i++) {
```

justin-tse commented on docs/chapter_computational_complexity/time_complexity.md:
```diff
@@ -220,13 +234,44 @@ $$
+    }
+    // 算法 C 时间复杂度：常数阶
+    function algorithm_C(n) {
+        for (var i = 0; i < 1000000; i++) {
```
```suggestion
        for (let i = 0; i < 1000000; i++) {
```

justin-tse commented on docs/chapter_computational_complexity/time_complexity.md:
```diff
@@ -220,13 +234,44 @@ $$
+    }
+    // 算法 B 时间复杂度：线性阶
+    function algorithm_B(n: number): void {
+        for (var i = 0; i < n; i++) {
```
```suggestion
        for (let i = 0; i < n; i++) {
```

justin-tse commented on docs/chapter_computational_complexity/time_complexity.md:
```diff
@@ -220,13 +234,44 @@ $$
+    }
+    // 算法 C 时间复杂度：常数阶
+    function algorithm_C(n: number): void {
+        for (var i = 0; i < 1000000; i++) {
```
```suggestion
        for (let i = 0; i < 1000000; i++) {
```

justin-tse commented on docs/chapter_computational_complexity/time_complexity.md:
```diff
@@ -354,20 +399,38 @@ $$
+        a += 1; // +1
+        a *= 2; // +1
+        // 循环 n 次
+        for(var i = 0; i < n; i++){ // +1
```
```suggestion
        for(let i = 0; i < n; i++){ // +1（每轮都执行 i ++）
```

justin-tse commented on docs/chapter_computational_complexity/time_complexity.md:
```diff
@@ -354,20 +399,38 @@ $$
+        a += 1; // +1
+        a *= 2; // +1
+        // 循环 n 次
+        for(var i = 0; i < n; i++){ // +1
```
```suggestion
        for(let i = 0; i < n; i++){ // +1（每轮都执行 i ++）
```

justin-tse commented on docs/chapter_computational_complexity/time_complexity.md:
````diff
@@ -530,13 +593,39 @@ $$
     ```js title=""
-
+    function algorithm(n) {
+        var a = 1;  // +0（技巧 1）
````
```suggestion
        let a = 1;  // +0（技巧 1）
```

justin-tse commented on docs/chapter_computational_complexity/time_complexity.md:
```diff
@@ -530,13 +593,39 @@ $$
+        var a = 1;  // +0（技巧 1）
+        a = a + n;  // +0（技巧 1）
+        // +n（技巧 2）
+        for (var i = 0; i < 5 * n + 1; i++) {
```
```suggestion
        for (let i = 0; i < 5 * n + 1; i++) {
```

justin-tse commented on docs/chapter_computational_complexity/time_complexity.md:
````diff
@@ -530,13 +593,39 @@ $$
     ```typescript title=""
-
+    function algorithm(n: number): void {
+        var a: number = 1;  // +0（技巧 1）
````
```suggestion
        let a = 1;  // +0（技巧 1）
```

justin-tse commented on docs/chapter_computational_complexity/time_complexity.md:
```diff
@@ -530,13 +593,39 @@ $$
+            console.log(0);
+        }
+        // +n*n（技巧 3）
+        for (var i = 0; i < 2 * n; i++) {
+            for (var j = 0; j < n + 1; j++) {
```
```suggestion
        for (let i = 0; i < 2 * n; i++) {
            for (let j = 0; j < n + 1; j++) {
```

justin-tse commented on docs/chapter_computational_complexity/time_complexity.md:
```diff
@@ -530,13 +593,39 @@ $$
+        var a: number = 1;  // +0（技巧 1）
+        a = a + n;  // +0（技巧 1）
+        // +n（技巧 2）
+        for (var i = 0; i < 5 * n + 1; i++) {
```
```suggestion
        for (let i = 0; i < 5 * n + 1; i++) {
```

justin-tse commented on docs/chapter_computational_complexity/time_complexity.md:
```diff
@@ -530,13 +593,39 @@ $$
+            console.log(0);
+        }
+        // +n*n（技巧 3）
+        for (var i = 0; i < 2 * n; i++) {
+            for (var j = 0; j < n + 1; j++) {
```
```suggestion
        for (let i = 0; i < 2 * n; i++) {
            for (let j = 0; j < n + 1; j++) {
```

justin-tse commented on docs/chapter_computational_complexity/time_complexity.md:
````diff
@@ -887,13 +998,29 @@ $$
     ```typescript title="time_complexity.ts"
-
+    /* 线性阶（遍历数组） */
+    function arrayTraversal(nums: number[]) {
````
```suggestion
    function arrayTraversal(nums: number[]): number {
```

justin-tse commented on docs/chapter_computational_complexity/time_complexity.md:
```diff
@@ -595,7 +684,7 @@ $$
 <div class="center-table" markdown>
 
-| 操作数量 $T(n)$         | 时间复杂度 $O(f(n))$   |
+| 操作数量 $T(n)$        | 时间复杂度 $O(f(n))$ |
```
```suggestion
| 操作数量 $T(n)$         | 时间复杂度 $O(f(n))$   |
```

justin-tse reviewed (changes requested):
Thank you @RiverTwilight for your clear code. It seems good and only has a small flaw.
And lack one part in the time_complexity.md:
=== "JavaScript"
    ```js title="worst_best_time_complexity.js"
    ```
=== "TypeScript"
    ```typescript title="worst_best_time_complexity.ts"
    ```
Please add it, thank you!

RiverTwilight commented:
Missing `worst_best_time_complexity.ts` added. Code run results here:

![image](https://user-images.githubusercontent.com/52880665/210687720-0f6df3e0-6bcf-495a-8e2a-46b738bfb737.png)

justin-tse commented on codes/javascript/chapter_computational_complexity/worst_best_time_complexity.js:
```diff
@@ -0,0 +1,45 @@
+    }
+    // 随机打乱数组元素
+    for (let i = 0; i < n; i++) {
+        let r = Math.floor(Math.random() * n);
```
```suggestion
        let r = Math.floor(Math.random() * (i + 1));
```

justin-tse commented on codes/javascript/chapter_computational_complexity/worst_best_time_complexity.js:
```diff
@@ -0,0 +1,45 @@
+/* 查找数组 nums 中数字 1 所在索引 */
+function findOne(nums) {
+    for (let i = 0; i < nums.length; i++) {
+        if (nums[i] == 1) {
```
```suggestion
        if (nums[i] === 1) {
```

justin-tse commented on codes/typescript/chapter_computational_complexity/worst_best_time_complexity.ts:
```diff
@@ -0,0 +1,45 @@
+    }
+    // 随机打乱数组元素
+    for (let i = 0; i < n; i++) {
+        let r = Math.floor(Math.random() * n);
```
```suggestion
        let r = Math.floor(Math.random() * (i + 1));
```

justin-tse commented on codes/typescript/chapter_computational_complexity/worst_best_time_complexity.ts:
```diff
@@ -0,0 +1,45 @@
+/* 查找数组 nums 中数字 1 所在索引 */
+function findOne(nums: number[]): number {
+    for (let i = 0; i < nums.length; i++) {
+        if (nums[i] == 1) {
```
```suggestion
        if (nums[i] === 1) {
```

justin-tse commented on docs/chapter_computational_complexity/time_complexity.md:
```diff
@@ -2049,13 +2328,89 @@ $$
+        }
+        // 随机打乱数组元素
+        for (let i = 0; i < n; i++) {
+            let r = Math.floor(Math.random() * n);
```
```suggestion
            let r = Math.floor(Math.random() * (i + 1));
```

justin-tse commented on docs/chapter_computational_complexity/time_complexity.md:
```diff
@@ -2049,13 +2328,89 @@ $$
+    /* 查找数组 nums 中数字 1 所在索引 */
+    function findOne(nums) {
+        for (let i = 0; i < nums.length; i++) {
+            if (nums[i] == 1) {
```
```suggestion
            if (nums[i] === 1) {
```

justin-tse commented on docs/chapter_computational_complexity/time_complexity.md:
```diff
@@ -2049,13 +2328,89 @@ $$
+        }
+        // 随机打乱数组元素
+        for (let i = 0; i < n; i++) {
+            let r = Math.floor(Math.random() * n);
```
```suggestion
            let r = Math.floor(Math.random() * (i + 1));
```

justin-tse commented on docs/chapter_computational_complexity/time_complexity.md:
```diff
@@ -2049,13 +2328,89 @@ $$
+    /* 查找数组 nums 中数字 1 所在索引 */
+    function findOne(nums: number[]): number {
+        for (let i = 0; i < nums.length; i++) {
+            if (nums[i] == 1) {
```
```suggestion
            if (nums[i] === 1) {
```

justin-tse reviewed (changes requested):
Thank you @RiverTwilight for your contribution!
The shuffle method has a little problem, which could not generate the true random data, you could check it in [Will It Shuffle?](https://bost.ocks.org/mike/shuffle/compare.html)
<img width="1027" alt="image" src="https://user-images.githubusercontent.com/24556310/210808080-b939392e-1ed9-434a-bd72-7211e1a2f345.png">
You could check one of the shuffle methods in [wiki](https://en.wikipedia.org/wiki/Fisher%E2%80%93Yates_shuffle#The_modern_algorithm)
Btw, I have updated the code, feel free to check it. Thank you.

RiverTwilight commented:
Thanks very much! @justin-tse  The code references in the doc have been updated.

justin-tse reviewed (approved):
@RiverTwilight Thank you for your contribution.
